### PR TITLE
Spectral Arc-Length Loss: FFT-weighted surface pressure loss

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Surface pressure along an airfoil is a 1D signal when parameterized by arc-length. Pointwise MAE equally weights errors at all positions. A spectral loss that upweights high-wavenumber components forces the model to reproduce sharp features — stagnation point, suction peak, pressure recovery slope — that dominate aerodynamic accuracy.

**Key distinction from the merged DCT loss (PR #2184):** That PR applied DCT on node-index order (arbitrary, physically meaningless). This PR sorts surface nodes by arc-length coordinate first, then applies torch.fft.rfft along that sorted dimension. Arc-length ordering makes the frequency spectrum physically meaningful: low frequencies = broad pressure trends, high frequencies = leading-edge singularity, stagnation point location, suction peak shape.

**Reference:** "Spectral Loss for Neural Operators" — NeurIPS ML4PS 2025 workshop. Also: Sobolev training for operator learning (arXiv:2402.09084, ICLR 2024).

## Instructions

Add a --spectral_arc_loss flag. When enabled, sort surface nodes by arc-length coordinate, then apply FFT and penalize high-frequency spectral errors.

### Step 1: Add arguments

```python
parser.add_argument('--spectral_arc_loss', action='store_true',
                    help='Spectral FFT loss on arc-length-sorted surface pressure nodes')
parser.add_argument('--spectral_arc_weight', type=float, default=0.1)
parser.add_argument('--spectral_arc_gamma', type=float, default=1.0,
                    help='Frequency weighting exponent (0.5=mild, 1=linear, 2=aggressive)')
```

### Step 2: Implement (add after surf_loss is computed)

```python
if args.spectral_arc_loss and surf_mask.any():
    B = pred.shape[0]
    spectral_losses = []
    for b in range(B):
        mask_b = surf_mask[b]
        if mask_b.sum() < 4:
            continue
        # saf_norm: arc-length normalized coord. Check which column it is in x features.
        # Try x[b, mask_b, 8] — adjust column if needed by inspecting feature names.
        arc_len = x[b, mask_b, 8]
        sort_idx = arc_len.argsort()
        pred_p = pred[b, mask_b, 2][sort_idx]  # pressure channel sorted by arc-length
        true_p = target[b, mask_b, 2][sort_idx]
        pred_fft = torch.fft.rfft(pred_p, norm='ortho')
        true_fft = torch.fft.rfft(true_p, norm='ortho')
        freqs = torch.arange(pred_fft.shape[0], device=pred.device, dtype=pred.dtype)
        weights = (freqs + 1.0) ** args.spectral_arc_gamma
        weights = weights / weights.mean()
        spectral_losses.append(((pred_fft - true_fft).abs() * weights).mean())
    if spectral_losses:
        surf_loss = surf_loss + args.spectral_arc_weight * torch.stack(spectral_losses).mean()
```

### Step 3: Run

Run 2 seeds with gamma=1.0, weight=0.1:

```bash
CUDA_VISIBLE_DEVICES=0 python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/spectral-arc-s42" --wandb_group "spectral-arc-length-loss" \
  --spectral_arc_loss --spectral_arc_weight 0.1 --spectral_arc_gamma 1.0 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling

CUDA_VISIBLE_DEVICES=1 python train.py --agent askeladd --seed 73 \
  --wandb_name "askeladd/spectral-arc-s73" --wandb_group "spectral-arc-length-loss" \
  --spectral_arc_loss --spectral_arc_weight 0.1 --spectral_arc_gamma 1.0 \
  [same flags as above]
```

Note: First verify which column index saf_norm corresponds to in x features. Print x.shape and inspect the feature ordering from the data pipeline if unsure.

## Baseline

- **p_in:** 11.742 | **p_oodc:** 7.643 | **p_tan:** 27.874 | **p_re:** 6.419
- Baseline W&B: k5qwvce4 (seed 42), 7oa5xfhi (seed 73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Spectral Arc-Length Loss (weight=0.1, gamma=1.0)

| Metric | Baseline | Seed 42 | Seed 73 | Avg | Δ vs baseline |
|--------|----------|---------|---------|-----|---------------|
| **p_in** | 11.742 | **11.203** | 12.017 | 11.610 | **-1.1%** |
| **p_oodc** | 7.643 | 7.767 | **7.717** | 7.742 | +1.3% |
| **p_tan** | 27.874 | 27.973 | 28.386 | 28.179 | +1.1% |
| **p_re** | 6.419 | **6.359** | 6.629 | 6.494 | +1.2% |

### Volume MAE (for reference)

| Metric | Seed 42 | Seed 73 |
|--------|---------|---------|
| vol_Ux (in-dist) | 0.492 | 0.507 |
| vol_Uy (in-dist) | 0.188 | 0.191 |
| vol_p (in-dist) | 12.467 | 12.221 |

### Run Details

- **Seed 42**: W&B run `z9fvhp17`, 144 epochs, ~75s/epoch, 46.1GB VRAM peak
- **Seed 73**: W&B run `ro4kwbp0`, 146 epochs, ~74s/epoch, 46.4GB VRAM peak
- **Commands used**:
  ```
  CUDA_VISIBLE_DEVICES=0 python train.py --agent askeladd --seed 42 \
    --wandb_name "askeladd/spectral-arc-s42" --wandb_group "spectral-arc-length-loss" \
    --spectral_arc_loss --spectral_arc_weight 0.1 --spectral_arc_gamma 1.0 \
    --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
    --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
    --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
    --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
    --pressure_first --pressure_deep \
    --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
    --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
    --gap_stagger_spatial_bias \
    --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
    --te_coord_frame --wake_deficit_feature --re_stratified_sampling
  ```
  (Same flags for seed 73 with `--seed 73` and `CUDA_VISIBLE_DEVICES=1`)

### Analysis

**Verdict: Neutral/marginal — no Pareto improvement over baseline.**

The spectral arc-length loss is essentially a wash. Seed 42 shows a small improvement on p_in (-4.6%) and p_re (-0.9%), but seed 73 is worse on both. Averaging across seeds, p_in improves marginally (-1.1%) while p_oodc, p_tan, and p_re all regress slightly (+1.1-1.3%).

**Why it didn't help much:**
1. **Redundancy with existing DCT freq loss**: The baseline already includes `--dct_freq_loss` which applies FFT-based frequency weighting (albeit on x-coordinate-sorted nodes rather than arc-length-sorted). The spectral arc-length loss adds a very similar signal — both penalize high-frequency spectral errors on surface pressure. The improvement from physically meaningful arc-length ordering appears too small to overcome the noise.
2. **Arc-length sorting via angle from centroid** is an approximation — it gives a closed-curve ordering but doesn't correspond to true arc-length parameterization (nodes are unevenly spaced along the foil). The physical meaningfulness advantage over x-coordinate sorting may be weaker than hypothesized.
3. **Weight=0.1 may be too high** alongside the existing DCT loss (weight=0.05), creating competing frequency-domain objectives that partially cancel each other out in the PCGrad 3-way setup.

### Suggested Follow-ups

1. **Try without DCT loss**: Run with `--spectral_arc_loss` but *without* `--dct_freq_loss` to isolate the effect and avoid redundancy. If the spectral arc loss alone matches or beats DCT loss alone, it validates the arc-length ordering hypothesis.
2. **Lower weight (0.02-0.05)**: The combined spectral+DCT signal may be too strong. Try weight=0.02 to see if a lighter touch helps.
3. **Per-foil arc-length from coordinates**: Instead of angle-from-centroid sorting, compute actual cumulative arc-length from consecutive node distances. This would give physically exact ordering.
4. **Complex-valued loss**: Currently using `.abs()` on the FFT difference (magnitude only). A loss on complex differences would also penalize phase errors, which correspond to spatial shifts of pressure features.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2320 because ADVISOR: Closing — spectral arc-length loss is essentially neutral (p_in -1.1% but p_oodc/p_tan/p_re all +1.1-1.3%). Net effect is a wash.

Your analysis is correct: the spectral loss is redundant with the existing DCT frequency loss. Both target the same spectral information, and the arc-length ordering advantage over x-coordinate sorting is too small to matter. The angle-from-centroid sorting is also only an approximation of true arc-length.

Takeaway: **our existing DCT freq loss already captures the spectral regularization benefit.** No further spectral loss variants are needed. Moving you to a fresh experiment..
